### DISCUSSION TITLE
bug: debugger does not start

### DISCUSSION THREAD

**Discussion by asfaltboy** [fredrikaverpil/neotest-golang#140] [Configuration]
### Did you check docs and existing issues?

- [X] I have read the [documentation](https://github.com/fredrikaverpil/neotest-golang/blob/main/README.md).
- [X] I have searched the existing [issues](https://github.com/fredrikaverpil/neotest-golang/issues).

### Neovim version (nvim -v)

0.10.0

### Operating system/version

macOS 14.5

### Describe the bug

neotest-golang runs a single test, but cannot start debugger for the same test.

Note: neotest-go starts the debugger (though it runs for all tests in the module 😢 ), but neotest-golang fails with an error:

```lua
neotest-golang: ...vchenko/.local/share/nvim/lazy/nvim-nio/lua/nio/init.lua:119: The coroutine failed with this message: 
...nvim/lazy/neotest/lua/neotest/client/strategies/init.lua:57: attempt to index local 'instance' (a nil value)
stack traceback:
	...nvim/lazy/neotest/lua/neotest/client/strategies/init.lua: in function 'run'
	...al/share/nvim/lazy/neotest/lua/neotest/client/runner.lua:130: in function '_run_spec'
	...al/share/nvim/lazy/neotest/lua/neotest/client/runner.lua:89: in function <...al/share/nvim/lazy/neotest/lua/neotest/client/runner.lua:88>
stack traceback:
	...ocal/share/nvim/lazy/neotest/lua/neotest/client/init.lua:89: in function <...ocal/share/nvim/lazy/neotest/lua/neotest/client/init.lua:88>
	[C]: in function 'error'
	...vchenko/.local/share/nvim/lazy/nvim-nio/lua/nio/init.lua:119: in function 'gather'
	...al/share/nvim/lazy/neotest/lua/neotest/client/runner.lua:94: in function '_run_tree'
	...al/share/nvim/lazy/neotest/lua/neotest/client/runner.lua:65: in function <...al/share/nvim/lazy/neotest/lua/neotest/client/runner.lua:22>
	[C]: in function 'xpcall'
	...ocal/share/nvim/lazy/neotest/lua/neotest/client/init.lua:84: in function 'run_tree'
	...al/share/nvim/lazy/neotest/lua/neotest/consumers/run.lua:85: in function 'func'
	...chenko/.local/share/nvim/lazy/nvim-nio/lua/nio/tasks.lua:173: in function <...chenko/.local/share/nvim/lazy/nvim-nio/lua/nio/tasks.lua:172>
```

### Steps To Reproduce

1. Start a debugger for a test case
2. An error occurs

### Expected Behavior

Should run the debugger

### Your Lua setup

```Lua
Using LazyVim I configure:


return {
  "nvim-neotest/neotest",
  opts = {
    adapters = {
      -- "neotest-go",
      "neotest-golang",
    },
  },
  dependencies = {
    {
      -- {
      --   "nvim-neotest/neotest-go",
      --   extra_args = { "-vet=off" }
      -- },
      {
        "fredrikaverpil/neotest-golang",
        args = { "-vet=off" }
      },
    },
  },
}
```

**Comment by fredrikaverpil** [comment]
Please provide a minimal reproducible Go test example.

**Comment by asfaltboy** [comment]
Check out https://github.com/golang-jwt/jwt , run `go build ./...` to make sure it builds, then run neovim in the dir and execute a debug test.

**Comment by fredrikaverpil** [comment]
I cloned down the repo, found a test at random and tried to both run it and then debug it. Works on my machine 😄 

<img width="899" alt="Screenshot 2024-07-15 at 15 08 18" src="https://github.com/user-attachments/assets/130624e7-80c1-4f38-8b9f-2d52489e602f">
<img width="1792" alt="Screenshot 2024-07-15 at 15 08 53" src="https://github.com/user-attachments/assets/ffafba44-4f28-4279-ac64-4f9ce88c88f8">

It feels like this could be a configuration issue on your end...?

Since [LazyVim is already using neotest-golang](https://github.com/LazyVim/LazyVim/blob/706ec4b6b6be4265cbcfd326d3216f2a29952b55/lua/lazyvim/plugins/extras/lang/go.lua#L128), I could quite easily try it out here too, and it works for me:

<img width="1395" alt="Screenshot 2024-07-15 at 15 14 44" src="https://github.com/user-attachments/assets/a5febc2a-f815-4454-8e8c-d6a847c316c6">

<img width="1792" alt="Screenshot 2024-07-15 at 15 15 01" src="https://github.com/user-attachments/assets/d10a1f19-7aaf-43f7-bb52-c0e10110ac7a">

**Comment by fredrikaverpil** [comment]
Were you able to run a test at all (not debug it)?

Could it be that you haven't enabled debugging in your neotest-golang config with `dap_go_enabled = true`?

What happens if you run `:checkhealth neotest-golang`?

**Reply by fredrikaverpil** [reply]
Instead of letting Neotest error like you observed, I'm now raising an error from neotest-golang (https://github.com/fredrikaverpil/neotest-golang/pull/141/commits/a17f3889564470839837ecbe8ad7d42d87c6a482). So if this was your issue, you should now see another error which is hopefully better att explaining what went wrong.

**Comment by fredrikaverpil** [comment]
I moved the thread here @asfaltboy as I suspect this to be a configuration issue.

**Comment by asfaltboy** [comment]
Thanks Fredrik, you're right I thought the config is identical to neotest-go, but didnd't realise I need to pass `dap_go_enabled = true` to enable debugging. This now works in another repo I have.

With regards to `jwt` it still doesn't work for me, but might be another issue.

I added it now to my config, and updated to latest release as of now. But sadly, the default package selection still fails for me in `jwt`, it emits: `"go", "test", "-json", "-v", "-race", "-count=1", "/Users/pavel.savchenko/Code/oss/jwt", "-run", "^TestECDSAVerify$"` which fails for me on the command line with this message:

```
directory ../../../Code/oss/jwt outside main module or its selected dependencies
```

**Reply by fredrikaverpil** [reply]
Hmmm. That's strange, as I could run (and debug) that test on my machine. Makes me doubt something is wrong in the jwt project. 

That capital C in "Code" caught my eye here (and maybe the "." in your username).

In your current setup, what happens if you run:

```bash
go test -v -race -count=1 /Users/pavel.savchenko/Code/oss/jwt -run ^TestECDSAVerify$
```

vs

```bash
go test -v -race -count=1 /Users/pavel.savchenko/code/oss/jwt -run ^TestECDSAVerify$
```

And also, what happens if you clone down the project into a folderpath which contains 1) no capital "C" and/or 2) no period sign and try to run/debug the test from there?

**Reply by asfaltboy** [reply]
I was able to run it in another path with a dot and capital, then I came back to my dir and it ran fine again... This must be something wrong with my machine, possibly anti-virus protection (this is a company machine, and I think it runs bit defender)

**Reply by fredrikaverpil** [reply]
I cloned down the project into my macOS user's `~/Public` folder and I'm getting a similar error that you had.

If I run the test with capital P it fails. If I run the test with lowercase p it passes. 🤷

**Comment by asfaltboy** [comment]
There may be more issues with directory/package lookup as well

For instance, I have a large repo, where the tested directory resolves to "./..." which runs, but way way slower than it does from CLI or with neotest-go.

I think the best performing selection for running a single test or a single file can be achieved by passing in the test file path or package directly, relative to the cwd, though it seems preventing it from using package internal names.

So for instance in jwt it would fail with a `[build failed]`:
```
go test  -v -race -count=1 -run '^TestECDSAVerify$' ./ecdsa_test.go
# command-line-arguments_test [command-line-arguments.test]
./ecdsa_test.go:141:4: undefined: benchmarkSigning
FAIL	command-line-arguments [build failed]
FAIL
```

The next best thing would be local dir `./.` in the case of `jwt`, but can be more deeply nested in larger repos.

And failing that, it would be the package name of course.

**Reply by fredrikaverpil** [reply]
In many cases, a Go project needs more than one file to build. Therefore it is unreliable to pass in a single filename to `go test`.  I actually didn't think it was possible to pass a filename at all. 

The adapter figures out the package name (using `go list`) and passes that in, instead of a filename. 

If you enable logging (instructions in the README) you can inspect the exact command that the adapter is generating, which might help track down issues here. Look for entries in the log file prefixed with `[neotest-golang]`. Feel free to raise a bug or enhancement issue with repro steps in case you have found something that isn't working as expected.

**Reply by fredrikaverpil** [reply]
So to be clear, the adapter defaults to `./...` for running all tests in a dir or file.

* https://github.com/fredrikaverpil/neotest-golang/blob/main/lua/neotest-golang/runspec/dir.lua#L30
* https://github.com/fredrikaverpil/neotest-golang/blob/main/lua/neotest-golang/runspec/file.lua#L29

It then tries to be more clever and figure out which package correlates to the dir or file. But if it fails, it will use `./...` which could be slower, yes.

You can run `go list` in the folder of your `go.mod` to see the output used by the adapter so it can translate a folderpath or file path into a package. It sounds like, in your case, this lookup is failing.

I think I will add some more logging today so all this will be easier to look into, by just reading the logs.

**Reply by asfaltboy** [reply]
Trouble is that it's so verbose on this repo, generating around 500k lines of debug rows per test run...

Hmm, with neotest-go it's not as verbose, I think it must not be doing as much work. It seems all it does is CD into the dir and run `go test ./` which seems to work fine for detecting a package. I wonder, would you be up for adding this naive approach as an alternative mechanism for directory/package detection,  perhaps behind an option flag?

**Reply by fredrikaverpil** [reply]
I'm not sure I understood your suggestion on a more naive package detection. Can you explain it in a separate [enhancement issue](https://github.com/fredrikaverpil/neotest-golang/issues/new?assignees=&labels=enhancement&projects=&template=feature.yml&title=feature%3A+) or in a PR?

**Reply by fredrikaverpil** [reply]
> Trouble is that it's so verbose on this repo, generating around 500k lines of debug rows per test run...

Do you mean the logfile gets that big?

**Reply by fredrikaverpil** [reply]
I added a little utility in this PR that will filter out non adapter related logs: https://github.com/fredrikaverpil/neotest-golang/pull/144

Give it a try if you wish, by setting `branch = "log-more"` in the adapter opts.

EDIT: now merged into `main`.
